interfaces/builtin/wayland.go

```diff
@@ -109,6 +109,12 @@ owner /run/user/[0-9]*/wayland-[0-9]* rw,
 /etc/drirc r,
 `
 
+const waylandConnectedPlugAppArmorCore = `
+# Allow access to the Wayland compositor server socket of the slot snap
```

Comment:
Should these perhaps be condensed to a single line:

```
owner {,/run/user/[0-9]*}/###SLOT_SECURITY_TAGS###/wayland-[0-9]* rw,
```

Reply:
I added the second rule to try and avoid the AppArmor denial mentioned in the previous comment. It looks like I'm still getting the same failure though. For example, on ubuntu-18.04-64:

```
audit[7974]: AVC apparmor="DENIED" operation="file_perm" profile="snap.wayland-client.wayland-client" name="/snap.wayland-server/wayland-9" pid=7974 comm="nc" requested_mask="r" denied_mask="r" fsuid=12345 ouid=0
audit[7974]: AVC apparmor="DENIED" operation="file_perm" profile="snap.wayland-client.wayland-client" name="/snap.wayland-server/wayland-9" pid=7974 comm="nc" requested_mask="r" denied_mask="r" fsuid=12345 ouid=0
```

I would have thought that'd match the second rule, but no such luck. The `ouid=0` bit seems a bit fishy though: the processes on both ends of the socket are running as the test user (uid 12345).

Reply:
@jrjohansen any ideas what could be going on here?

Reply:
Only guesses. Can we get an attached full copy of the generated profile? Yes I can and will generate one but its possible it will come out different.

Reply:
Here are the two policies from the test snaps using the snap build from this PR, with the wayland interface connected between the two:

[snap.wayland-client.wayland-client.txt](https://github.com/snapcore/snapd/files/9502612/snap.wayland-client.wayland-client.txt)
[snap.wayland-server.wayland-server.txt](https://github.com/snapcore/snapd/files/9502614/snap.wayland-server.wayland-server.txt)

Reply:
so this is failing because the socket file /snap.wayland-server/wayland-9 is owned by root (ouid=0), and the access rule only allows access if the socket/file is owned by the task (uid=12345).

either the socket/file needs to be chowned or the rule needs to drop the owner conditional changing to

```
 /{,run/user/[0-9]*/}###SLOT_SECURITY_TAGS###/wayland-[0-9]* rw,
```

Reply:
I don't think that's the case. In the test, the listening socket is created when we run the wayland-server snap as the `test` user:

    tests.session -u test exec wayland-server &

That creates `/run/user/12345/snap.wayland-server/wayland-9`. I then hard link that socket to `/run/user/12345/wayland-9` with:

    ln "$USER_RUNTIME_DIR"/snap.wayland-server/wayland-9 "$USER_RUNTIME_DIR"/wayland-9

The ln command is running as root, but the ownership is tied to the inode and we're just creating a new name for the socket inode created by `test`.

Finally, I'm invoking the client in the same way the server was, so I'd expect uids to match there too:

    tests.session -u test exec wayland-client

I'll try adding some more debug output to the spread test to verify socket ownership, and try running `ln` as the test user too to see if that helps.

Reply:
It is possible there is an apparmor bug, but the log message very much says that apparmor thinks the socket is owned by root.

```
  fsuid=12345 ouid=0
```

fsuid is the tasks current fsuid, ouid should be the inode's uid

Reply:
This is from the most recent CI run for ubuntu-18.04-64. The socket is definitely owned by the test user:

```
+ ls -li /run/user/12345/snap.wayland-server/wayland-9 /run/user/12345/wayland-9
19 srwxr-xr-x 2 test test 0 Sep  7 06:42 /run/user/12345/snap.wayland-server/wayland-9
19 srwxr-xr-x 2 test test 0 Sep  7 06:42 /run/user/12345/wayland-9
```

But the denial lists `ouid=0`:

```
Sep 07 06:42:43 sep070543-644342 audit[14140]: AVC apparmor="DENIED" operation="file_perm" profile="snap.wayland-client.wayland-client" name="/snap.wayland-server/wayland-9" pid=14140 comm="nc" requested_mask="r" denied_mask="r" fsuid=12345 ouid=0
Sep 07 06:42:43 sep070543-644342 audit[14140]: AVC apparmor="DENIED" operation="file_perm" profile="snap.wayland-client.wayland-client" name="/snap.wayland-server/wayland-9" pid=14140 comm="nc" requested_mask="r" denied_mask="r" fsuid=12345 ouid=0
```

It also seems a bit weird that the path in the denial message is listed as `/snap.wayland-server/wayland-9`. I'd usually expect a full path, but this looks like it is relative to the root of the tmpfs mounted at `$XDG_RUNTIME_DIR`.

Reply:
The only reason I can think of for the tmpfs would not have the full path would be because the tmpfs is not part of the tasks mount namespace. It is being forcibly attached via the attach_disconnected flag on the profile. Removing attach_disconnected should result in DENIED messages with info="Failed name lookup - disconnected path"

Reply:
Also, I have found the bug in apparmor around the object uid for unix sockets

Reply:
The same tmpfs would be available in both snaps mount namespaces. This is the one created by systemd and mounted at `/run/user/12345`, rather than something snapd created. If the tmpfs wasn't available in the mount namespace, I wouldn't expect the connect() call to progress far enough to trigger the denial too.

I'll add that I can accept "the AppArmor in these old releases is buggy" as an answer. The test seems to be passing reliably on 22.04 / Core 22, which are the main cases I care about here.

Reply:
attach_disconnected would apply to the connect. The mount namespace the object is opened in matters, so having the same inode in different mount namespaces can very much result in disconnected paths. With unix sockets each end of the socket is considered and that is likely where we are seeing the disconnected path from.